_source/_posts/2022-04-11-java-cli-device-grant.adoc

```diff
@@ -0,0 +1,412 @@
+
+To keep things all contained in a single `java` file, I'll use https://www.jbang.dev[JBang]. If you haven't used JBang before, the TL;DR it makes it easy to create Java... scripts. Your project dependencies get added to the top of the `java` file.
+
+Create a new `DeviceGrant.java` and add `jackson-databind` as a dependency:
+
+====
+[source,java]
+----
+//DEPS com.fasterxml.jackson.core:jackson-databind:2.13.2 // <.>
+
+import com.fasterxml.jackson.databind.ObjectMapper; // <.>
+
+import java.awt.*; // <.>
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.net.URI;
+import java.net.URLEncoder;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.Map;
+
+import static java.util.stream.Collectors.joining;
+
+public class DeviceGrant {
+
+    public static void main(String[] args) throws Exception {
+
+    }
+}
+----
+
+<.> Make sure there is no space between the `//` and `DEPS`.
+<.> Add the following imports, they will be used below.
+<.> AWT? Yes, keep reading I'll get to that 😉
+====
+
+Execute the application from the command line using:
+
+[source,shell]
+----
+jbang DeviceGrant.java
+----
+
+This application doesn't do anything yet, we can fix the in the next few steps!
+
+=== A little more boilerplate
+
+The fluent `HttpClient` is handy, but along with not having JSON support, it doesn't have direct support for form encoded requests. Fortunately, both are easy to fix by adding a couple of helper methods!
+
+In `DeviceGrant.java` add the following two methods:
+
+[source,java]
+----
+// Convert a map into a encoded form body
+private static HttpRequest.BodyPublisher formBody(Map<String, String> params) {
+    // Wrap an encoded String in a BodyPublisher
+    return HttpRequest.BodyPublishers.ofString(
+        // url encode <key>=<value>&
+        params.entrySet().stream()
+            .map(entry -> URLEncoder.encode(entry.getKey(), UTF_8)
+                          + "="
+                          + URLEncoder.encode(entry.getValue(), UTF_8))
+            .collect(joining("&")));
+}
+
+// parse a response body using Jackson
+public static <W> HttpResponse.BodySubscriber<W> parseJson(Class<W> targetType) {
+    return HttpResponse.BodySubscribers.mapping(
+        HttpResponse.BodySubscribers.ofByteArray(),
+        (byte[] bytes) -> {
+            try {
+                ObjectMapper objectMapper = new ObjectMapper();
+                return objectMapper.readValue(bytes, targetType);
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        });
+}
+----
+
+== Initiate the authorization request
+
+From the above link:#setup-okta[Setup Okta] section you have a public "client id" and "issuer" URI, for convenience's sake, set those as `String` variables:
```

Comment:
```suggestion
From the above link:#setup-okta[Set up Okta] section you have a public "client id" and "issuer" URI. For the sake of convenience, set those as `String` variables:
```
that apostrophe wasn't incorrect, but it's easier to read without it imo
do you need to adjust the anchor name since "set up" is now two words?